crates/models/falcon/src/lib.rs

```diff
@@ -163,18 +183,35 @@ impl KnownModel for Falcon {
                 current = ctx0.op_norm(&input_layer);
                 current = ctx0.op_add(
                     &ctx0.op_mul(
-                        &ctx0.op_repeat(&self.layers[il].attention_norm, &current),
+                        &ctx0.op_repeat(&self.layers[il].input_layernorm, &current),
                         &current,
                     ),
-                    &ctx0.op_repeat(&self.layers[il].attention_norm_b, &current),
+                    &ctx0.op_repeat(&self.layers[il].input_layernorm_b, &current),
                 );
 
                 layernorm_output = current.share();
 
+                // Falcon-40B only
+                if n_head_kv != 1 {
+                    current = ctx0.op_add(
+                        &ctx0.op_mul(
+                            &ctx0.op_repeat(
+                                self.layers[il].attention_norm.as_ref().unwrap(),
+                                &current,
+                            ),
+                            &current,
+                        ),
+                        &ctx0.op_repeat(
+                            self.layers[il].attention_norm_b.as_ref().unwrap(),
+                            &current,
+                        ),
+                    );
+                }
+
                 // compute QKV
                 current = ctx0.op_mul_mat(&self.layers[il].query_key_value, &current);
 
-                let fused_qkv_row_nb = (n_embd + 2 * (n_embd / n_head)) * f32_size;
+                let fused_qkv_row_nb = head_dim * (n_head + 2 * n_head_kv) * f32_size;
```

Comment:
Is this from the pytorch implementation? Or where did you reference this from?

Reply:
The code references [ggllm.cpp](https://github.com/cmp-nct/ggllm.cpp/blob/60f82cac75f4f3be4a85dba397c0632a28de28ba/libfalcon.cpp#L2195), and it should be equal to the value of `current.get_nb()[1]`.